src/lib/crypto/kimchi_backend/gadgets/foreign_field.ml

```diff
@@ -1395,40 +1397,66 @@ let mul (type f) (module Circuit : Snark_intf.Run with type field = f)
         ; basic =
             Kimchi_backend_common.Plonk_constraint_system.Plonk_constraint.T
               (ForeignFieldMul
-                 { (* Current row *) left_input0
+                 { (* left input *)
+                   left_input0
                  ; left_input1
                  ; left_input2
-                 ; right_input0
+                 ; (* right input *) right_input0
                  ; right_input1
                  ; right_input2
-                 ; carry1_lo
-                 ; carry1_hi
-                 ; carry0
-                 ; quotient0
+                 ; (* remainder *) remainder01
+                 ; remainder2
+                 ; (* quotient *) quotient0
                  ; quotient1
                  ; quotient2
-                 ; quotient_bound_carry
-                 ; product1_hi_1
-                 ; (* Next row *) remainder0
-                 ; remainder1
-                 ; remainder2
-                 ; quotient_bound01
-                 ; quotient_bound2
-                 ; product1_lo
+                 ; quotient_hi_bound
+                 ; (* products *) product1_lo
                  ; product1_hi_0
-                 ; (* Coefficients *) foreign_field_modulus0
-                 ; foreign_field_modulus1
-                 ; foreign_field_modulus2
+                 ; product1_hi_1
+                 ; (* carries *) carry0
+                 ; carry1_0
+                 ; carry1_12
+                 ; carry1_24
+                 ; carry1_36
+                 ; carry1_48
+                 ; carry1_60
+                 ; carry1_72
+                 ; carry1_84
+                 ; carry1_86
+                 ; carry1_88
+                 ; carry1_90
+                 ; (* Coefficients *) foreign_field_modulus2
                  ; neg_foreign_field_modulus0
                  ; neg_foreign_field_modulus1
                  ; neg_foreign_field_modulus2
                  } )
         } ) ;
+
+  (*
+   * Add external checks (and related)
+   *)
+  External_checks.append_multi_range_check external_checks
+    (quotient0, quotient1, quotient2) ;
+
+  External_checks.append_multi_range_check external_checks
+    (quotient_hi_bound, product1_lo, product1_hi_0) ;
+
+  (* Instead of appending external check for compact MRC for remainder,
+   * this is added directly, so that the standard limbs
+   * (remainder0, remainder1, remainder2) are copyable in witness cells *)
+  let remainder0, remainder1 =
+    Range_check.compact_multi (module Circuit) remainder01 remainder2
+  in
+
+  External_checks.append_bound_check external_checks ~do_multi_range_check:false
```

Comment:
In the RFC I suggested to make the RCs as well as the high limb bounds check on the remainder optional. The idea was that in some cases, you don't want those checks:
* If you assert that the remainder equals a known value, you can do that even from the compact representation, and you don't even need the compact RC. An example would be in foreign field inversion, where you check that r = 1
* If the remainder doesn't become an input to another ffmul, or if you want to add a full bounds check on it, you don't need the high limb bounds check.

I'd prefer if we exposed the possibility to make those optimizations, and add a flag for skipping any RCs on the remainder (default value: false)

Reply:
I've already restored this functionality here -> https://github.com/MinaProtocol/mina/pull/13725